Gems/Atom/RPI/Code/Source/RPI.Public/Material/MaterialShaderParameter.cpp

```diff
@@ -0,0 +1,373 @@
+
+        void MaterialShaderParameter::SetStructuredBufferData(
+            const MaterialShaderParameterDescriptor* desc, const AZStd::span<const uint8_t>& deviceData, const int deviceIndex)
+        {
+            auto& binding{ desc->m_structuredBufferBinding };
+
+            AZ_Assert(
+                binding.m_elementSize * binding.m_elementCount == deviceData.size_bytes(),
+                "Size mismatch when setting the Material Shader Parameter data for %s %s: expected: %lld bytes, provided %lld bytes",
+                desc->m_typeName.c_str(),
+                desc->m_name.c_str(),
+                binding.m_elementSize * binding.m_elementCount,
+                deviceData.size_bytes());
+
+            size_t minBufferSize = binding.m_offset + binding.m_elementSize * binding.m_elementCount;
+            if (m_structuredBufferData[deviceIndex].size() < minBufferSize)
+            {
+                m_structuredBufferData[deviceIndex].resize(minBufferSize, 0);
+            }
+            for (int byteIndex = 0; byteIndex < deviceData.size(); ++byteIndex)
+            {
+                m_structuredBufferData[deviceIndex][binding.m_offset + byteIndex] = deviceData[byteIndex];
+            }
+        }
+
+        AZStd::span<const uint8_t> MaterialShaderParameter::GetRawBufferParameterData(
+            const MaterialShaderParameterDescriptor* desc, const uint32_t deviceIndex) const
+        {
+            AZ_Assert(desc, "GetRawBufferParameterData: desc is nullptr!");
+            if (desc)
+            {
+                auto offset = desc->m_structuredBufferBinding.m_offset;
+                auto size = desc->m_structuredBufferBinding.m_elementSize * desc->m_structuredBufferBinding.m_elementCount;
+                for (auto& [device, structuredBufferData] : m_structuredBufferData)
+                {
+                    if (device == deviceIndex)
+                    {
+                        return AZStd::span<const uint8_t>{ &structuredBufferData[offset], size };
+                    }
+                }
+            }
+            return {};
+        }
+
+        bool MaterialShaderParameter::SetMaterialSrgDeviceReadIndex(
+            const MaterialShaderParameterDescriptor* desc, [[maybe_unused]] const int deviceIndex, const int32_t readIndex)
+        {
+            if (!m_shaderResourceGroup)
+            {
+                return false;
+            }
+            auto* index = AZStd::get_if<RHI::ShaderInputConstantIndex>(&desc->m_srgInputIndex);
+            if (index && index->IsValid())
+            {
+                // TODO: this doesn't work, i can't set different constant values in different device-SRGs (yet)
+                // m_shaderResourceGroup->GetRHIShaderResourceGroup()
+                //     ->GetDeviceShaderResourceGroup(deviceIndex)
+                //     ->GetData()
+                //     ->SetConstant(index, readIndex);
+                return m_shaderResourceGroup->SetConstant(*index, readIndex);
```

Comment:
I assume this means that single material SRGs will not work with MultiGPU. Ideally this could be fixed in this PR, or alternatively create an issue for this.

Reply:
created issue #18494